spring-web/src/main/java/org/springframework/web/context/support/WebApplicationContextUtils.java

```diff
@@ -125,6 +136,63 @@ public static WebApplicationContext getWebApplicationContext(ServletContext sc,
 		return (WebApplicationContext) attr;
 	}
 
+	/**
+	 * Registers a custom {@link WebApplicationContext} for this web app.
+	 * @param sc ServletContext to find the web application context for
+	 * @param attrName the name of the ServletContext attribute to use
+	 * @param context the desired WebApplicationContext to store
+	 * @since 4.2
+	 */
+	public static void registerWebApplicationContext(ServletContext sc, String attrName, WebApplicationContext context) {
+		Assert.notNull(sc, "ServletContext must not be null");
+		Assert.notNull(attrName, "attrName must not be null");
+		Assert.notNull(context, "WebApplicationContext must not be null");
+
+		sc.setAttribute(attrName, context);
+
+		@SuppressWarnings("unchecked")
+		LinkedHashSet<String> attrNames = (LinkedHashSet<String>) sc.getAttribute(APPLICATION_CONTEXT_ATTRIBUTES_ATTRIBUTE);
+		if(attrNames == null) {
+			attrNames = new LinkedHashSet<String>();
+		}
+
+		attrNames.add(attrName);
+		sc.setAttribute(APPLICATION_CONTEXT_ATTRIBUTES_ATTRIBUTE, attrNames);
+	}
+
+	/**
+	 * Gets all the known {@link WebApplicationContext} for this web app. The order is the
+	 * root context (if specified) followed by any context registered using
+	 * registerWebApplicationContext in the order they were registered.
+	 *
+	 * @param sc ServletContext to find the web application context for. Cannot be null.
+	 * @return the known {@link WebApplicationContext} for this web app. Cannot be null,
+	 * but may be empty.
+	 * @since 4.2
+	 */
+	public static List<WebApplicationContext> getRegisteredWebApplicationContext(ServletContext sc) {
```

Comment:
This method should be named `getRegisteredWebApplicationContexts` (i.e. plural).